plugins/coordinator/plugin.go

```diff
@@ -141,36 +166,135 @@ func run(plugin *node.Plugin) {
+					log.Warn(err)
+					continue
+				}
+				lastCheckpointIndexNonLazy++
+				lastCheckpointHashNonLazy = checkpointHash
+
+			case <-nextCheckpointSemiLazySignal:
+				// check the thresholds again, because a new milestone could have been issued in the meantime
+				if trackedTailsCount := selectorSemiLazy.GetTrackedTailsCount(); trackedTailsCount < maxTrackedTails {
+					continue
 				}
 
+				tips, err := selectorSemiLazy.SelectTips(0)
+				if err != nil {
+					// issuing checkpoint failed => not critical
+					if err != mselection.ErrNoTipsAvailable {
+						log.Warn(err)
+					}
+					continue
+				}
+
+				// issue a semi-lazy checkpoint
+				checkpointHash, err := coo.IssueCheckpoint("semi-lazy", lastCheckpointIndexSemiLazy, lastCheckpointHashSemiLazy, tips)
+				if err != nil {
+					// issuing checkpoint failed => not critical
+					log.Warn(err)
+					continue
+				}
+				lastCheckpointIndexSemiLazy++
+				lastCheckpointHashSemiLazy = checkpointHash
+
 			case <-nextMilestoneSignal:
-				err, criticalErr := coo.IssueMilestone()
+
+				// issue a new semi-lazy checkpoint right in front of the milestone
+				semiLazyTips, err := selectorSemiLazy.SelectTips(1)
+				if err != nil {
+					// issuing checkpoint failed => not critical
+					if err != mselection.ErrNoTipsAvailable {
+						log.Warn(err)
+					}
+				} else {
+					semiLazyCheckpointHash, err := coo.IssueCheckpoint("semi-lazy", lastCheckpointIndexSemiLazy, lastCheckpointHashSemiLazy, semiLazyTips)
+					if err != nil {
+						// issuing checkpoint failed => not critical
+						log.Warn(err)
+					} else {
+						// use the new checkpoint hash
+						lastCheckpointHashSemiLazy = semiLazyCheckpointHash
+					}
+				}
+
+				// issue a new non-lazy checkpoint right in front of the milestone
+				nonLazyTips, err := selectorNonLazy.SelectTips(1)
+				if err != nil {
+					// issuing checkpoint failed => not critical
+					if err != mselection.ErrNoTipsAvailable {
+						log.Warn(err)
+					}
+				} else {
+					// add the semi-lazy checkpoint to the list the non-lazy checkpoint should reference
+					nonLazyTips = append(nonLazyTips, lastCheckpointHashSemiLazy)
+
+					nonLazyCheckpointHash, err := coo.IssueCheckpoint("non-lazy", lastCheckpointIndexNonLazy, lastCheckpointHashNonLazy, nonLazyTips)
+					if err != nil {
+						// issuing checkpoint failed => not critical
+						log.Warn(err)
+					} else {
+						// use the new checkpoint hash
+						lastCheckpointHashNonLazy = nonLazyCheckpointHash
+					}
+				}
+
+				milestoneHash, err, criticalErr := coo.IssueMilestone(lastMilestoneHash, lastCheckpointHashNonLazy)
 				if criticalErr != nil {
 					log.Panic(criticalErr)
 				}
 				if err != nil {
+					if err == tangle.ErrNodeNotSynced {
+						// Coordinator is not synchronized, trigger the solidifier manually
+						tangleplugin.TriggerSolidifier()
+					}
 					log.Warn(err)
 				}
 
+				// remember the last milestone hash
+				lastMilestoneHash = milestoneHash
```

Comment:
if you had an error when issuing the milestone, which value is inside milestoneHash?
Maybe break inside the error check?